src/plugins/crypto/gcrypt_operations.c

```diff
@@ -18,41 +20,126 @@
-	if (key == NULL)
+	gcry_error_t gcry_err;
+	kdb_octet_t salt[ELEKTRA_CRYPTO_DEFAULT_SALT_LEN];
+	kdb_octet_t keyBuffer[KEY_BUFFER_SIZE];
+
+	// generate the salt
+	gcry_create_nonce (salt, sizeof (salt));
+	elektraCryptoNormalizeRandomString (salt, sizeof (salt));
+	keySetMeta (k, ELEKTRA_CRYPTO_META_SALT, (char *)salt);
+
+	// read iteration count
+	const kdb_unsigned_long_t iterations = elektraCryptoGetIterationCount (config);
+
+	// receive master password from the configuration
+	Key * master = ksLookupByName (config, ELEKTRA_CRYPTO_PARAM_MASTER_PWD, 0);
+	if (!master)
+	{
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, errorKey, "missing %s", ELEKTRA_CRYPTO_PARAM_MASTER_PWD);
+		return -1;
+	}
+	Key * msg = keyDup (master);
+	if (elektraCryptoGpgDecryptMasterPassword (config, errorKey, msg) != 1)
+	{
+		goto error;
+	}
+
+	// generate/derive the cryptographic key and the IV
+	if ((gcry_err = gcry_kdf_derive (keyValue (msg), keyGetValueSize (msg), GCRY_KDF_PBKDF2, GCRY_MD_SHA512, salt, sizeof (salt),
+					 iterations, KEY_BUFFER_SIZE, keyBuffer)))
 	{
-		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, errorKey, "missing %s in configuration",
-				    ELEKTRA_CRYPTO_PARAM_KEY_PATH);
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_INTERNAL_ERROR, errorKey, "PBKDF2 failed because: %s", gcry_strerror (gcry_err));
+		goto error;
 	}
-	return key;
+
+	keySetBinary (cKey, keyBuffer, ELEKTRA_CRYPTO_GCRY_KEYSIZE);
+	keySetBinary (cIv, keyBuffer + ELEKTRA_CRYPTO_GCRY_KEYSIZE, ELEKTRA_CRYPTO_GCRY_BLOCKSIZE);
+
+	keyDel (msg);
+	return 1;
+
+error:
+	keyDel (msg);
+	return -1;
 }
 
 /**
- * @brief read the cryptographic initialization vector (IV) from the given keyset.
- * @retval NULL on error
- * @retval address of the (Elektra) key in the given keyset holding the IV to be used.
+ * @brief derive the cryptographic key and IV for a given (Elektra) Key k
+ * @param config KeySet holding the plugin/backend configuration
+ * @param errorKey holds an error description in case of failure
+ * @param k the (Elektra)-Key to be encrypted
+ * @param cKey (Elektra)-Key holding the cryptographic material
+ * @param cIv (Elektra)-Key holding the initialization vector
+ * @retval -1 on failure. errorKey holds the error description.
+ * @retval 1 on success
  */
-static Key * elektraCryptoReadParamIv (KeySet * config, Key * errorKey)
+static int getKeyIvForDecryption (KeySet * config, Key * errorKey, Key * k, Key * cKey, Key * cIv)
 {
-	Key * iv = ksLookupByName (config, ELEKTRA_CRYPTO_PARAM_IV_PATH, 0);
-	if (iv == NULL)
+	gcry_error_t gcry_err;
+	kdb_octet_t keyBuffer[KEY_BUFFER_SIZE];
+
+	// get the salt
+	const Key * salt = keyGetMeta (k, ELEKTRA_CRYPTO_META_SALT);
+	if (!salt)
 	{
-		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, errorKey, "missing %s in configuration",
-				    ELEKTRA_CRYPTO_PARAM_IV_PATH);
+		ELEKTRA_SET_ERRORF (ELEKTRA_ERROR_CRYPTO_CONFIG_FAULT, errorKey, "missing salt as meta-key %s for key %s",
```

Comment:
For now I will stick to "metakey" as it is the most common usage. We may change it later on.